### ISSUE TITLE
Rethink CI command with an empahsis on very helpful diagnostics

### ISSUE THREAD

**Issue by cowboyd** [thefrontside/bigtest#430]
Right now, our error handling is competent, but not great. For example, if we have a failed assertion in this test suite:

```ts
export default test("Simulated GraphQL API")
  .step("start GraphQL simulator", async () => ({
    simulator: await startBrowserSimulation()
  }))
  .step("create two companies", async ({simulator}) => {
    let [first, second] = simulator.createList('company', 2);

    return { first, second };
  })
  .step("fetch companies over graphql", async () => {
    let query = `{ companies { id name } }`;
    return {
      data: await request('/graphql', query)
    };
  })
  .assertion("it returns two companies", async ({ data } ) => {
    await assertEqual(data.companies.length, 3);
  })
  .assertion("the returned companies are serialized properly", async ({ data, first, second }) => {
    await assertEqual(data.companies[0].name, first.name);
    await assertEqual(data.companies[1].name, second.name);
  })

```

The output upon an assertion failure looks like this:

```
$ yarn bigtest test
yarn run v1.22.4
✓ [step]       Simulated GraphQL API -> start GraphQL simulator
✓ [step]       Simulated GraphQL API -> create two companies
✓ [step]       Simulated GraphQL API -> fetch companies over graphql
⨯ [assertion]  Simulated GraphQL API -> it returns two companies
|   AssertionError: expected 3 == 2
|       at _callee$ (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:66693:23)
|       at tryCatch (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:196:23)
|       at Generator.invoke [as _invoke] (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:413:28)
|       at Generator.next (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:249:27)
|       at asyncGeneratorStep (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:861:28)
|       at _next (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:883:13)
|       at http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:890:11
|       at new Promise (<anonymous>)
|       at http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:879:16
|       at _assertEqual (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:66702:27)
✓ [assertion]  Simulated GraphQL API -> the returned companies are serialized properly

⨯ FAILURE finished in 0.21s
Steps:         3 ok     0 failed     0 disregarded
Assertions:    1 ok     1 failed     0 disregarded
```

The most important information is that "expected two companies to be in the list, but there were three" but this gets completely occluded by a stacktrace that only an effection implementer and chrome inspector could love. There is a sequence of operations that resulted in a failure of expectation, but that sequence has very little to do with the sequence in which a tick of javascript functions were called. At best, the stack trace is 3rd order information behind the console logs from the agent, and the actual sequence of actions that were run.

When a failure happens, we want to see primarily what was going on

```
1. start graphql simulator
2. create two companies
3. fetch all companies over graphql
4. expected there to be three companies in the returned data, but there were 2
```

We don't need to see that stack trace at all unless there is set a `--show-javascript-stacktrace` method or a `--vv` for "very verbose" setting.

In fact, we may not even want to see the console log statements without an `--agent-log` option.

Proposal for CI streaming UI
-----------------------------

> Note this is not for a "smart run" UI where we can do things like have progress bars. It is also not fore a stateful CLI

After attempting a spike with @rwwagner90 this week, I believe the simplest and most helpful API will be to treat the individual steps as just a '.' progress, with maybe an 'x' for failure and then print a good summary of what happened. Let's use the following test tree:

```ts
export default test("Simulated GraphQL API")
  .step("start GraphQL simulator", async () => ({
    simulator: await startBrowserSimulation()
  }))
  .child("Company API", test => test
    .step("create two companies", async ({simulator}) => {
      let [first, second] = simulator.createList('company', 2);
      return { first, second };
    })
    .step("fetch companies over graphql", async () => {
      let query = `{ companies { id name } }`;
      return {
        data: await request('/graphql', query)
      };
     })
    .assertion("it returns two companies", async ({ data } ) => {
       await assertEqual(data.companies.length, 2);
     })
    .assertion("the returned companies are serialized properly", async ({ data, first, second }) => {
       await assertEqual(data.companies[0].name, first.name);
       await assertEqual(data.companies[1].name, second.name);
   }))
  .child("Person API", test => test
    .step("create two people", async ({simulator}) => {
      let [first, second] = simulator.createList('person', 2);
      return { first, second };
    })
    .step("fetch people over graphql", async () => {
      let query = `{ people { id name } }`;
      return {
        data: await request('/graphql', query)
      };
     })
   .assertion("it returns two people", async ({ data } ) => {
       await assertEqual(data.people.length, 2);
     })
   .assertion("the returned people are serialized properly", async ({ data, first, second }) => {
       await assertEqual(data.people[0].name, first.name);
       await assertEqual(data.people[1].name, second.name);
   }));
 ```

A passing test run would look like:

```
$ bigtest test
..........
✅ Passed (3/3)
```

If we passed the `--show-tree` flag, then it would render the entire test tree, and look like:

```
$ bigtest test --show-tree
..........
🧪Simulated GraphQL
  ↪ Start GraphQL Simulator
  🧪Company API
    ↪ create two companies
    ↪ fetch all companies over graphql
    ✅ it returns two companies
    ✅ the returned companies serialize properly
  🧪Person API
    ↪ create two people
    ↪ fetch all people over graphql
    ✅ it returns two people
    ✅ the returned people serialize properly

✅ Pass (3/3)
```

Let's assume that  _people_ test branch had an assertion failure. We would only show the failing test case in sequence

```
$ bigtest test
.......x..

🧪Simulated GraphQL API
    ↪ Start GraphQL Simulator
  🧪Person API
    ↪ create two people
    ↪ fetch all people over graphql
    🛑 the returned people serialize properly
        expected "name" to be defined, but it was undefined.
        at graphql-test.ts:35:2

✅ Pass (2/3)
❗️ Failed (1/3)
```
Notice how all of the steps are included in the failure as a singular sequence even though the "start graphql simulator" is a step in the parent.

With the `--show-tree` flag, we would render the whole tree.
```
$ bigtest test --show-tree
..........

🧪Simulated GraphQL
  ↪ Start GraphQL Simulator
  🧪Company API
    ↪ create two companies
    ↪ fetch all companies over graphql
    ✅ it returns two companies
    ✅ the returned companies serialize properly
  🧪Person API
    ↪ create two people
    ↪ fetch all people over graphql
    ✅ it returns two people
    🛑 the returned people serialize properly
      expected "name" to be defined, but it was undefined.
      at graphql-test.ts:35:2

✅ Pass (2/3)
❗️ Failed (1/3)
```

When a step errors, we show the test case that it failed on. let's say we had an error in the creating two companies:

```
$ bigtest test
......

🧪Simulated GraphQL > Company API
  ↪ Start GraphQL Simulator
  💥 create two companies
      TypeError: simulation.create is not a function, it is undefined.
      at simulator.js:45:3

✅ Pass (2/3)
❗️ Failed (1/3)
```

Notice that the successive steps are not even rendered because they were not attempted, but with the `show-tree` we'll show how they were skipped.

```
$ bigtest test --show-tree
......

🧪Simulated GraphQL
  ↪ Start GraphQL Simulator
  🧪Company API
    💥 create two companies
        TypeError: simulation.create is not a function, it is undefined.
        at simulator.js:45:3
    🔘 fetch all companies over graphql
    🔘it returns two companies
    🔘 the returned companies serialize properly
  🧪Person API
    ↪ create two people
    ↪ fetch all people over graphql
    ✅ it returns two people
    ✅ the returned people serialize properly

❗️ Failed (1/2)
```

It will be handly to "inline" console output that was emitted during the course of a step, because if something failed, in the pipeline, then we'll want to know when it happened. So let's say that we get a console warning while we are starting the GraphQL simulator:

```
$ bigtest test
......

🧪Simulated GraphQL > Company API
  ↪ Start GraphQL Simulator
    ⚠️ WARNING: you have two instances of the graphql package running at the same time
      this can cause undefined results when initiating queries.
  💥 create two companies
      TypeError: simulation.create is not a function, it is undefined.
      at simulator.js:45:3
```

This gives a clear indicator that even though starting the graphql simulator was successful, something was not quite right about the startup and we should investigate further.

What about the case where the parent step fails. In that case, both test cases will be discarded:

```
$ bigtest test
.

🧪Simulated GraphQL
  💥 Start GraphQL Simulator
    TypeError: module not found "graphql"

🔘 Discarded (2/3)
❗️ Failed (1/3)
```

**Comment by jorgelainfiesta** [comment]
During the milestones meeting of today, we agreed to focus first on the CI output. I think the ideas for CLI presented in this issue are applicable to CI, given that the interaction is relatively limited and there's no coloring involved.

**Comment by jgnieuwhof** [comment]
Just a small note! It might be nice to choose icons that all share the same shape, or style.

**Comment by jnicklas** [comment]
Though there is more work to do, an initial version of this was implemented in #462
